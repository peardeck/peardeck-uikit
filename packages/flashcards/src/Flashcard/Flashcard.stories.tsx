import React from "react";
import { Flashcard } from "./Flashcard";

export default {
  component: Flashcard,
  title: "Flashcards|Flashcard",
};

export const Default = () => <Flashcard />;
