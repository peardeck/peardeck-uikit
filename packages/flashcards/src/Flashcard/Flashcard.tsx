import React from "react";
import styled from "styled-components";
import { Box, Card } from "@peardeck-uikit/core";

export interface FlashcardProps {
  flipped: boolean;
  slide: boolean;
  speed: number;
}

const Container = styled(Box)`
  perspective: 1000px;
  height: 200px;
  width: 200px;
`;

const Flipper = styled(Box)<FlashcardProps>`
  // Adjust speed here.
  ${({ speed }): string => `transition: ${speed}s`};
  transform-style: preserve-3d;
  position: relative;

  ${({ flipped, slide }): string =>
    flipped && slide
      ? `
    transform-origin: center right;
    transform: translateX(-100%) rotateY(180deg);`
      : ""};

  ${({ flipped, slide }): string =>
    flipped && !slide ? "transform: rotateY(180deg);" : ""};
`;

const Face = styled(Card)`
  backface-visibility: hidden;

  // height: 200px;
  // width: 200px;
`;

const Front = styled(Card)`
  z-index: 2;
  /* for firefox 31 */
  transform: rotateY(0deg);
  background: red;
`;

const Back = styled(Face)`
  z-index: 1;
  transform: rotateY(180deg);
  background: blue;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
`;

export const Flashcard = ({
  flipped = false,
  slide = false,
  speed = 0.5,
}: FlashcardProps): JSX.Element => (
  <Container>
    <Flipper flipped={flipped} slide={slide} speed={speed}>
      <Front>This is the front</Front>
      <Back>This is the back</Back>
    </Flipper>
  </Container>
);
