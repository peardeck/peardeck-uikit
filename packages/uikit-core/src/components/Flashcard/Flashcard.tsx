import React from "react";
import { Box, BoxProps } from "../Box/Box";
import { Card } from "../Card/Card";
import { Text } from "../Text/Text";
import { FlippableBox } from "../FlippableBox/FlippableBox";
import { Heading1 } from "../Heading/Heading";

export interface FlashcardProps {
  image?: React.ReactNode;
  term: React.ReactNode;
  example?: React.ReactNode;
  definition?: React.ReactNode;
  height: number;
  width: number;
  flipped?: boolean;
  flipSpeed?: number | string;
  slideAnimation?: boolean;

  // TODO: Have to figure out how to do with this color
  // property type mismatch
  color?: string;
}

export const Flashcard = ({
  image,
  term,
  example,
  definition,
  height,
  width,
  flipped = false,
  flipSpeed,
  slideAnimation = false,
  ...rest
}: FlashcardProps & BoxProps): JSX.Element => (
  <FlippableBox
    {...rest}
    height={height}
    width={width}
    slideAnimation={slideAnimation}
    front={
      <Card
        height={height}
        width={width}
        p={0}
        bg="pear"
        display="flex"
        alignItems="stretch"
      >
        {image && (
          <Box
            display="flex"
            flex={1}
            m={2}
            p={2}
            border="1px solid #698600"
            justifyContent="center"
            alignItems="center"
          >
            {image}
          </Box>
        )}
        <Box m={2} flex={2} overflow="auto">
          {definition && <Heading1 pb={2}>{definition}</Heading1>}
          {example && <Text pt={2}>{example}</Text>}
        </Box>
      </Card>
    }
    back={
      <Card
        height={height}
        width={width}
        p={0}
        bg="#eee"
        display="flex"
        alignItems="center"
        justifyContent="center"
      >
        <Heading1>{term}</Heading1>
      </Card>
    }
    flipped={flipped}
    flipSpeed={flipSpeed}
  />
);
