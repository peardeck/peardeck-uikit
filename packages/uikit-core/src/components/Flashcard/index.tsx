export { Flashcard, FlashcardProps } from "./Flashcard";
