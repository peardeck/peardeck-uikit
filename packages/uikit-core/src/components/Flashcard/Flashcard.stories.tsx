import React, { ElementType } from "react";
import { boolean, number, text, select } from "@storybook/addon-knobs";
import {
  IconAlphabetical,
  IconAccount,
  IconCalendar,
  IconCheckmark,
  IconClassroom,
  IconImage,
} from "@peardeck/uikit-icons";
import { colorKnobs, spaceKnobs } from "../../theme/knobs";
import { Box } from "../Box/Box";
import { Flashcard } from "./Flashcard";

export default {
  component: Flashcard,
  title: "core|Flashcard",
};

const ICONS: { [key: string]: ElementType } = {
  IconAlphabetical,
  IconAccount,
  IconCalendar,
  IconCheckmark,
  IconClassroom,
  IconImage,
};
const ICONS_KEYS = Object.keys(ICONS);

const KNOB_GROUP = "Flashcard";

export const Default = () => {
  const height = number("Container Height", 300, {}, KNOB_GROUP);
  const width = number("Container Width", 500, {}, KNOB_GROUP);
  const slideAnimation = boolean("Use Slide Animation", false, KNOB_GROUP);
  const iconKey = select("Icon", ICONS_KEYS, ICONS_KEYS[1], KNOB_GROUP);
  const IconComponent = ICONS[iconKey];

  return (
    <Box {...colorKnobs()} {...spaceKnobs()} m={3}>
      <Flashcard
        p={4}
        height={height}
        width={width}
        slideAnimation={slideAnimation}
        image={IconComponent && <IconComponent height="100%" width="100%" />}
        term={text("Term", "Figurative Language", KNOB_GROUP)}
        definition={text(
          "Definition",
          "Language that uses imagination or exaggeration and does not stick to the facts",
          KNOB_GROUP
        )}
        example={text(
          "Example",
          "I had to walk 15 miles to school in the snow, uphill, in bare feet.",
          KNOB_GROUP
        )}
        flipped={boolean("Flipped", false, KNOB_GROUP)}
        flipSpeed={`${number("Flip Speed", 0.5, {}, KNOB_GROUP)}s`}
      />
    </Box>
  );
};
