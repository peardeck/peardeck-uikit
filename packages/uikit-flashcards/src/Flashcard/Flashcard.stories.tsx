import React, { useState } from "react";
import { boolean, number } from "@storybook/addon-knobs";
import { Button } from "@peardeck/uikit-core";
import { Flashcard } from "./Flashcard";

export default {
  component: Flashcard,
  title: "Flashcards|Flashcard",
};

export const Default = () => {
  const [isFlipped, setFlipped] = useState(false);
  return (
    <>
      <Button buttonSize="small" mb={2} onClick={() => setFlipped(!isFlipped)}>
        Flip It
      </Button>
      <Flashcard
        flipped={isFlipped}
        slide={boolean("Slide animation", false)}
        speed={number("Speed", 0.5)}
      />
    </>
  );
};
