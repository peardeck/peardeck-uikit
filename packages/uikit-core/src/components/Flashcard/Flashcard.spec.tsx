import { create } from "react-test-renderer";
import React from "react";
import { Flashcard } from "./Flashcard";

describe("Flashcard", () => {
  test("it matches the snapshot", () => {
    const json = create(
      <Flashcard height={300} width={100} term="" definition="" example="" />
    ).toJSON();
    expect(json).toMatchSnapshot();
  });

  describe("flipped", () => {
    test("it matches the snapshot", () => {
      const json = create(
        <Flashcard
          height={500}
          width={300}
          term="Claim"
          definition="What you are trying to prove, your thesis"
          example="Aliens Exist"
          flipped
        />
      ).toJSON();
      expect(json).toMatchSnapshot();
    });
  });
});
