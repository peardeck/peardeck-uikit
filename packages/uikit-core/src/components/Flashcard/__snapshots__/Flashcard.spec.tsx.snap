// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flashcard flipped it matches the snapshot 1`] = `
.c4 {
  overflow: auto;
  margin: 8px;
  -webkit-flex: 2;
  -ms-flex: 2;
  flex: 2;
}

.c3 {
  border-radius: 3px;
  background-color: #c9de55;
  height: 500px;
  width: 300px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  box-shadow: 0 1px 4px 0 rgba(0,0,0,0.27);
  border-radius: 3px;
}

.c8 {
  border-radius: 3px;
  background-color: #eee;
  height: 500px;
  width: 300px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  box-shadow: 0 1px 4px 0 rgba(0,0,0,0.27);
  border-radius: 3px;
}

.c6 {
  padding-top: 8px;
}

.c0 {
  height: 500px;
  width: 300px;
  -webkit-perspective: 1000px;
  -moz-perspective: 1000px;
  -ms-perspective: 1000px;
  perspective: 1000px;
}

.c1 {
  -webkit-transition-duration: 0.5s;
  transition-duration: 0.5s;
  -webkit-transform-style: preserve-3d;
  -ms-transform-style: preserve-3d;
  transform-style: preserve-3d;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  -webkit-transform: rotateY(180deg);
  -ms-transform: rotateY(180deg);
  transform: rotateY(180deg);
}

.c2 {
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  overflow: auto;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 2;
  -webkit-transform: rotateY(0deg);
  -ms-transform: rotateY(0deg);
  transform: rotateY(0deg);
}

.c7 {
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  overflow: auto;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 1;
  -webkit-transform: rotateY(180deg);
  -ms-transform: rotateY(180deg);
  transform: rotateY(180deg);
}

.c5 {
  padding-bottom: 8px;
}

@media print {
  .c0 {
    -webkit-break-after: always;
    break-after: always;
  }
}

@media print {
  .c1 {
    -webkit-transform-origin: unset;
    -ms-transform-origin: unset;
    transform-origin: unset;
    -webkit-transition-duration: unset;
    transition-duration: unset;
    -webkit-transform: unset;
    -ms-transform: unset;
    transform: unset;
    position: unset;
    top: unset;
    bottom: unset;
    left: unset;
    right: unset;
  }
}

@media print {
  .c2 {
    -webkit-backface-visibility: unset;
    backface-visibility: unset;
    overflow: unset;
    position: unset;
    top: unset;
    bottom: unset;
    left: unset;
    right: unset;
    box-shadow: none;
    border-radius: 0;
    border: 1px dotted;
    border-bottom: 1px solid;
  }
}

@media print {
  .c2 {
    -webkit-transform: unset;
    -ms-transform: unset;
    transform: unset;
  }
}

@media print {
  .c7 {
    -webkit-backface-visibility: unset;
    backface-visibility: unset;
    overflow: unset;
    position: unset;
    top: unset;
    bottom: unset;
    left: unset;
    right: unset;
    box-shadow: none;
    border-radius: 0;
    border: 1px dotted;
    border-bottom: 1px solid;
  }
}

@media print {
  .c7 {
    -webkit-transform: rotate(180deg);
    -ms-transform: rotate(180deg);
    transform: rotate(180deg);
  }
}

<div
  className="c0"
  height={500}
  width={300}
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
        display="flex"
        height={500}
        width={300}
      >
        <div
          className="c4"
          overflow="auto"
        >
          <h1
            className="c5"
          >
            What you are trying to prove, your thesis
          </h1>
          <span
            className="c6"
          >
            Aliens Exist
          </span>
        </div>
      </div>
    </div>
    <div
      className="c7"
    >
      <div
        className="c8"
        display="flex"
        height={500}
        width={300}
      >
        <h1
          className=""
        >
          Claim
        </h1>
      </div>
    </div>
  </div>
</div>
`;

exports[`Flashcard it matches the snapshot 1`] = `
.c4 {
  overflow: auto;
  margin: 8px;
  -webkit-flex: 2;
  -ms-flex: 2;
  flex: 2;
}

.c3 {
  border-radius: 3px;
  background-color: #c9de55;
  height: 300px;
  width: 100px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  box-shadow: 0 1px 4px 0 rgba(0,0,0,0.27);
  border-radius: 3px;
}

.c6 {
  border-radius: 3px;
  background-color: #eee;
  height: 300px;
  width: 100px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  box-shadow: 0 1px 4px 0 rgba(0,0,0,0.27);
  border-radius: 3px;
}

.c0 {
  height: 300px;
  width: 100px;
  -webkit-perspective: 1000px;
  -moz-perspective: 1000px;
  -ms-perspective: 1000px;
  perspective: 1000px;
}

.c1 {
  -webkit-transition-duration: 0.5s;
  transition-duration: 0.5s;
  -webkit-transform-style: preserve-3d;
  -ms-transform-style: preserve-3d;
  transform-style: preserve-3d;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
}

.c2 {
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  overflow: auto;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 2;
  -webkit-transform: rotateY(0deg);
  -ms-transform: rotateY(0deg);
  transform: rotateY(0deg);
}

.c5 {
  -webkit-backface-visibility: hidden;
  backface-visibility: hidden;
  overflow: auto;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 1;
  -webkit-transform: rotateY(180deg);
  -ms-transform: rotateY(180deg);
  transform: rotateY(180deg);
}

@media print {
  .c0 {
    -webkit-break-after: always;
    break-after: always;
  }
}

@media print {
  .c1 {
    -webkit-transform-origin: unset;
    -ms-transform-origin: unset;
    transform-origin: unset;
    -webkit-transition-duration: unset;
    transition-duration: unset;
    -webkit-transform: unset;
    -ms-transform: unset;
    transform: unset;
    position: unset;
    top: unset;
    bottom: unset;
    left: unset;
    right: unset;
  }
}

@media print {
  .c2 {
    -webkit-backface-visibility: unset;
    backface-visibility: unset;
    overflow: unset;
    position: unset;
    top: unset;
    bottom: unset;
    left: unset;
    right: unset;
    box-shadow: none;
    border-radius: 0;
    border: 1px dotted;
    border-bottom: 1px solid;
  }
}

@media print {
  .c2 {
    -webkit-transform: unset;
    -ms-transform: unset;
    transform: unset;
  }
}

@media print {
  .c5 {
    -webkit-backface-visibility: unset;
    backface-visibility: unset;
    overflow: unset;
    position: unset;
    top: unset;
    bottom: unset;
    left: unset;
    right: unset;
    box-shadow: none;
    border-radius: 0;
    border: 1px dotted;
    border-bottom: 1px solid;
  }
}

@media print {
  .c5 {
    -webkit-transform: rotate(180deg);
    -ms-transform: rotate(180deg);
    transform: rotate(180deg);
  }
}

<div
  className="c0"
  height={300}
  width={100}
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
        display="flex"
        height={300}
        width={100}
      >
        <div
          className="c4"
          overflow="auto"
        >
          
          
        </div>
      </div>
    </div>
    <div
      className="c5"
    >
      <div
        className="c6"
        display="flex"
        height={300}
        width={100}
      >
        <h1
          className=""
        >
          
        </h1>
      </div>
    </div>
  </div>
</div>
`;
